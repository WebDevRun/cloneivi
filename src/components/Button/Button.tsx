import React from 'react'
import cn from 'classnames'
import styles from './Button.module.scss'
import Image from 'next/image'

interface ButtonProps {
  /**
   * Назначение кнопки
   */
  mode?:
    | 'primary'
    | 'secondary'
    | 'player'
    | 'trailer'
    | 'share'
    | 'sign_in'
    | 'search'
    | 'defer'
    | 'free'
    | 'rate_sm'
    | 'rate_md'
    | 'filter'
  /**
   * Иконка
   */
  iconSrc?: string
  /**
   * Alt текст
   */
  iconAlt?: string
  /**
   * Основной текст на кнопке
   */
  text?: string
  /**
   * Дополнительный текст на кнопке
   */
  subText?: string
  /**
   * Обработчик нажатия
   */
  onClick?: () => void
}

/**
 * Primary UI component
 */

export const Button = ({
  mode = 'secondary',
  text,
  subText,
  iconSrc = '',
  iconAlt = '',
}: ButtonProps) => {
  const mainCn = cn(styles.button, styles[mode])

  return (
    <button type="button" className={mainCn}>
      {iconSrc && iconAlt && (
        <div className={styles.icon}>
          <Image src={iconSrc} alt={iconAlt} />
        </div>
      )}
      {text && (
        <div className={styles.main_text}>
          {text}
          <div className={styles.sub_text}>{subText}</div>
        </div>
      )}
    </button>
  )
}
