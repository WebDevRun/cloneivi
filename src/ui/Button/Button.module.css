.button {
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 0.5rem;
  transition-duration: 0.4s;
  transition-property: color, fill, border-color, background-color;
}
.button img {
  display: flex;
  align-items: center;
}
.button .icon {
  display: flex;
  justify-content: center;
  align-items: center;
  margin-right: 0.5rem;
  transition-duration: 0.4s;
}

.active svg {
  fill: #fff;
}

.passive svg {
  fill: rgba(255, 255, 255, 0.72);
}

.active {
  color: #fff;
}

.passive {
  color: rgba(255, 255, 255, 0.72);
}

.transparent {
  background-color: transparent;
  color: rgba(255, 255, 255, 0.72);
}
.transparent svg {
  fill: rgba(255, 255, 255, 0.72);
}
.transparent:hover {
  color: #fff;
}
.transparent:hover .icon svg {
  fill: #fff;
}

.red {
  background-color: #ea003d;
  color: #fff;
}
.red:hover {
  background-color: #ff0f4d;
}

.gray {
  background-color: #3a3352;
  color: #fff;
}
.gray:hover {
  background-color: #494069;
}

.primary {
  background-image: linear-gradient(270deg, #ea003d 0, #c447ff 100%);
}

.small {
  height: 2rem;
  padding: 0.4375rem 0.6875rem;
}
.small .mainText {
  font-size: 0.8125rem;
  line-height: 1rem;
  font-weight: 500;
}

.middle {
  height: 2.5rem;
  padding: 0.5625rem 0.9375rem;
  font-size: 0.9375rem;
  line-height: 1.25rem;
  font-weight: 500;
}

.big {
  height: 3rem;
  padding: 0.3125rem 0.9375rem;
  font-size: 0.9375rem;
  line-height: 1.25rem;
  font-weight: 500;
  width: 20rem;
}

.borderSm,
.borderMd {
  background: transparent;
  border-radius: 0.375rem;
  padding: 0.1875rem 0.4375rem;
  border: 1px solid rgba(255, 255, 255, 0.32);
}
.borderSm .mainText,
.borderMd .mainText {
  font-size: 0.8125rem;
  line-height: 1rem;
  font-weight: 500;
}
.borderSm:hover,
.borderMd:hover {
  border: 1px solid white;
  background: transparent;
}
.borderSm:not(:hover),
.borderMd:not(:hover) {
  transition-duration: 0.4s;
  transition-property: border;
}

.borderSm {
  height: 1.5rem;
}

.borderMd {
  height: 2rem;
}

.borderBg {
  height: 3rem;
  border-radius: 0.5rem;
  padding: 0.75rem;
  border: solid 0.125rem rgba(255, 255, 255, 0.48);
}
.borderBg:hover {
  border: solid 0.125rem #fff;
  box-shadow: 0 0 0.9375rem rgba(255, 255, 255, 0.4);
}
.borderBg:hover svg {
  fill: #fff;
}

.full {
  width: 100%;
}

.onlyIcon .icon {
  margin-right: 0;
}
.onlyIcon.small {
  width: 2rem;
}
.onlyIcon.middle {
  width: 2.5rem;
}
.onlyIcon.big {
  width: 3rem;
}

.noneFields {
  margin: 0;
  padding: 0;
  height: auto;
}

.rating {
  height: 3.5rem;
  width: 3.5rem;
  padding: 0;
  background-color: rgba(255, 255, 255, 0.16);
  border-radius: 0.75rem;
}
.rating > .mainText {
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: #73a32f;
  height: 2.75rem;
  width: 2.75rem;
  border-radius: 0.5rem;
  font-size: 0.9375rem;
  line-height: 1.25rem;
  font-weight: 500;
}
.rating .subText {
  display: none;
}
.rating:hover {
  background-color: rgba(255, 255, 255, 0.29);
}/*# sourceMappingURL=Button.module.css.map */