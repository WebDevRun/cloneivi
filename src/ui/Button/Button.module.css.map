{"version":3,"sources":["Button.module.scss","Button.module.css"],"names":[],"mappings":"AAMA;EACE,aAAA;EACA,uBAAA;EACA,mBAAA;EACA,qBAAA;EACA,yBAAA;EACA,gEAAA;ACLF;ADOE;EACE,aAAA;EACA,mBAAA;ACLJ;ADQE;EACE,aAAA;EACA,uBAAA;EACA,mBAAA;EACA,oBAAA;EACA,yBAAA;ACNJ;;ADUA;EACE,UA3BM;ACoBR;;ADUA;EACE,+BAAA;ACPF;;ADUA;EACE,WAnCM;AC4BR;;ADUA;EACE,gCAAA;ACPF;;ADUA;EACE,6BAAA;EACA,gCAAA;ACPF;ADSE;EACE,+BAAA;ACPJ;ADUE;EACE,WAnDI;AC2CR;ADUI;EACE,UAtDE;AC8CR;;ADaA;EACE,yBAAA;EACA,WA7DM;ACmDR;ADYE;EACE,yBAAA;ACVJ;;ADcA;EACE,yBAAA;EACA,WAtEM;AC2DR;ADaE;EACE,yBAAA;ACXJ;;ADeA;EACE,kEAAA;ACZF;;ADeA;EACE,YAAA;EACA,4BAAA;ACZF;ADcE;EACE,oBAAA;EACA,iBAAA;EACA,gBAAA;ACZJ;;ADgBA;EACE,cAAA;EACA,4BAAA;EACA,oBAAA;EACA,oBAAA;EACA,gBAAA;ACbF;;ADgBA;EACE,YAAA;EACA,4BAAA;EAEA,oBAAA;EACA,oBAAA;EACA,gBAAA;EACA,YAAA;ACdF;;ADiBA;;EAEE,uBAAA;EACA,uBAAA;EACA,4BAAA;EACA,2CAAA;ACdF;ADgBE;;EACE,oBAAA;EACA,iBAAA;EACA,gBAAA;ACbJ;ADgBE;;EACE,uBAAA;EACA,uBAAA;ACbJ;ADgBE;;EACE,yBAAA;EACA,2BAAA;ACbJ;;ADiBA;EACE,cAAA;ACdF;;ADiBA;EACE,YAAA;ACdF;;ADiBA;EACE,YAAA;EACA,qBAAA;EACA,gBAAA;EACA,gDAAA;ACdF;ADgBE;EACE,2BAAA;EACA,kDAAA;ACdJ;ADgBI;EACE,UAzJE;AC2IR;;ADmBA;EACE,WAAA;AChBF;;ADoBE;EACE,eAAA;ACjBJ;ADoBE;EACE,WAAA;AClBJ;ADqBE;EACE,aAAA;ACnBJ;ADsBE;EACE,WAAA;ACpBJ;;ADwBA;EACE,SAAA;EACA,UAAA;EACA,YAAA;ACrBF;;ADwBA;EACE,cAAA;EACA,aAAA;EACA,UAAA;EACA,2CAAA;EACA,sBAAA;ACrBF;ADuBE;EACE,aAAA;EACA,uBAAA;EACA,mBAAA;EACA,yBAAA;EACA,eAAA;EACA,cAAA;EACA,qBAAA;EACA,oBAAA;EACA,oBAAA;EACA,gBAAA;ACrBJ;ADwBE;EACE,aAAA;ACtBJ;ADyBE;EACE,2CAAA;ACvBJ","file":"Button.module.css"}