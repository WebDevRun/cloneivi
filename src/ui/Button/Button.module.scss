@use '../../styles/functions' as fn;
$white: #fff;
$alpha48: 0.48;
$alpha72: 0.72;

.button {
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: fn.toRem(8px);
  transition-duration: 0.4s;

  img {
    display: flex;
    align-items: center;
  }

  .icon {
    display: flex;
    justify-content: center;
    align-items: center;
    margin-right: fn.toRem(8px);
    transition-duration: 0.4s;
  }
}

.active svg {
  fill: $white;
}

.passive svg {
  fill: rgba($white, $alpha72);
}

.active {
  color: $white;
}

.passive {
  color: rgba($white, $alpha72);
}

.transparent {
  background-color: transparent;
  color: rgba($white, $alpha72);

  svg {
    fill: rgba($white, $alpha72);
  }

  &:hover {
    color: $white;
    transition-duration: 0.4s;

    svg {
      fill: $white;
      transition-duration: 0.4s;
    }
  }
}

.red {
  background-color: #ea003d;
  color: $white;

  &:hover {
    background-color: #ff0f4d;
  }
}

.gray {
  background-color: #3a3352;
  color: $white;

  &:hover {
    background-color: #494069;
  }
}

.primary {
  background-image: linear-gradient(270deg, #ea003d 0, #c447ff 100%);
}

.small {
  height: fn.toRem(32px);
  padding: fn.toRem(7px) fn.toRem(11px);

  .mainText {
    font-size: fn.toRem(13px);
    line-height: fn.toRem(16px);
    font-weight: 500;
  }
}

.middle {
  height: fn.toRem(40px);
  padding: fn.toRem(9px) fn.toRem(15px);
  font-size: fn.toRem(15px);
  line-height: fn.toRem(20px);
  font-weight: 500;
}

.big {
  height: fn.toRem(48px);
  padding: fn.toRem(5px) fn.toRem(15px);

  font-size: fn.toRem(15px);
  line-height: fn.toRem(20px);
  font-weight: 500;
  width: fn.toRem(320px);

  transition-property: border-color, background-color;
}

.borderSm,
.borderMd {
  background: transparent;
  border-radius: fn.toRem(6px);
  padding: fn.toRem(3px) fn.toRem(7px);
  border: 1px solid rgba($white, 0.32);

  .mainText {
    font-size: fn.toRem(13px);
    line-height: fn.toRem(16px);
    font-weight: 500;
  }

  &:hover {
    border: 1px solid rgba($white, 1);
    background: transparent;
  }

  &:not(:hover) {
    transition-duration: 0.4s;
    transition-property: border;
  }
}

.borderSm {
  height: fn.toRem(24px);
}

.borderMd {
  height: fn.toRem(32px);
}

.borderBg {
  height: fn.toRem(48px);
  border-radius: fn.toRem(8px);
  padding: fn.toRem(12px);
  border: solid fn.toRem(2px) rgba($white, 0.48);

  &:hover {
    border: solid fn.toRem(2px) $white;
    box-shadow: 0 0 fn.toRem(15px) rgba($white, 0.4);

    svg {
      fill: $white;
    }
  }
}

.onlyIcon {
  .icon {
    margin-right: 0;
  }

  &.small {
    width: fn.toRem(32px);
  }

  &.middle {
    width: fn.toRem(40px);
  }

  &.big {
    width: fn.toRem(48px);
  }
}

.noneFields {
  margin: 0;
  padding: 0;
  height: auto;
}
