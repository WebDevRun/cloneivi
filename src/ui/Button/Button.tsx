import cn from 'classnames'
import React, { FC, MouseEventHandler } from 'react'

import { INameIcons } from '@/types/Icons'

import { Svg } from '../Svg'

import styles from './Button.module.scss'

export interface ButtonProps {
  icon?: INameIcons
  background?: 'gray' | 'primary' | 'red' | 'transparent'
  theme?: 'active' | 'passive'
  withBorder?: 'borderNone' | 'borderSm' | 'borderMd' | 'borderBg'
  size?: 'small' | 'middle' | 'big'
  text?: string
  subText?: string
  onClick?: MouseEventHandler<HTMLButtonElement>
}

export const Button: FC<ButtonProps> = ({
  icon,
  background = 'red',
  theme = 'active',
  withBorder = 'borderNone',
  size = 'middle',
  text,
  subText,
  onClick,
}) => {
  const mainCn = cn(
    styles.button,
    styles[background],
    styles[theme],
    styles[withBorder],
    styles[size],
    !text && styles.onlyIcon,
  )

  const iconSize = size === 'big' ? 'big' : 'middle'

  return (
    <button type='button' className={mainCn} onClick={onClick}>
      {icon && (
        <div className={styles.icon}>
          {<Svg icon={icon} size={iconSize} />}
        </div>
      )}
      {text && (
        <div className={styles.mainText}>
          {text}
          <div className={styles.subText}>{subText}</div>
        </div>
      )}
    </button>
  )
}
