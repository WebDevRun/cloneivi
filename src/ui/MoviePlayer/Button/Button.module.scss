@use '@styles/functions' as fn;
@use '../general' as gn;

.button {
  @include gn.buttonStyles;
  @include gn.playingButtonStyles;
}

.image {
  @include gn.buttonImageStyles;
}
