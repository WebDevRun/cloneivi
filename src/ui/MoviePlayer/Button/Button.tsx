import Image from 'next/image'
import { FC } from 'react'

import setting from '@assets/images/player/setting.svg'
import subtitles from '@assets/images/player/subtitles.svg'

import styles from './Button.module.scss'

export interface ButtonProps {
  image: 'setting' | 'subtitles'
}

const images = {
  setting,
  subtitles,
}

export const Button: FC<ButtonProps> = ({ image }) => {
  const setImages = (image: ButtonProps['image']) => images[image]

  return (
    <button className={styles.button}>
      <Image className={styles.image} src={setImages(image)} alt={`${image}`} />
    </button>
  )
}
