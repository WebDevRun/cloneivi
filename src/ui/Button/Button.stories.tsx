import { Meta, StoryObj } from '@storybook/react'

import { Button } from './Button'
import '@styles/index.scss'

const meta: Meta<typeof Button> = {
  title: 'Button',
  component: Button,
  tags: ['autodocs'],
  argTypes: {
    size: {
      control: { type: 'select' },
      description: 'Размер',
    },
    icon: {
      control: { type: 'select' },
      description: 'Название иконки',
    },
    background: {
      control: { type: 'select' },
      description: 'Фон',
    },
    withBorder: {
      control: { type: 'select' },
      description: 'Рамка',
    },
    text: {
      control: 'text',
      description: 'Основной текст на кнопке',
    },
    subText: {
      type: 'string',
      description: 'Дополнительный текст на кнопке. Не работает без основного текста',
    },
    onClick: {
      type: 'function',
      description: 'Обработчик нажатия',
    },
  },

  parameters: {
    backgrounds: {
      default: 'twitter',
      values: [
        { name: 'twitter', value: '#00aced' },
        { name: 'facebook', value: '#3b5998' },
      ],
    },
  },
}

export default meta
type Story = StoryObj<typeof Button>

export const Player: Story = {
  name: 'С двойным текстом',
  args: {
    size: 'big',
    text: 'Смотреть с рекламой',
    subText: 'серия 1 сезон 1',
  },
}

export const PrimaryMob: Story = {
  name: 'Для мобильных устройств',
  args: {
    size: 'small',
    background: 'gray',
    text: 'Смотреть 30 дней за 1₽',
  },
}

export const Primary: Story = {
  args: {
    background: 'primary',
    text: 'Оплатить подписку',
  },
}

export const Secondary: Story = {
  args: {
    size: 'middle',
    text: 'Смотреть по подписке',
  },
}

export const Trailer: Story = {
  name: 'Серая с иконкой',
  args: {
    icon: 'search',
    background: 'gray',
    text: 'Трейлер',
  },
}

export const Search: Story = {
  name: 'Прозрачная с иконкой',
  args: {
    background: 'transparent',
    icon: 'search',
    text: 'Поиск',
  },
}

export const BorderSm: Story = {
  name: 'Прозрачная с рамкой маленькая',
  args: {
    withBorder: 'borderSm',
    text: 'Оценить',
  },
}

export const RateMd: Story = {
  name: 'Прозрачная с рамкой средняя',
  args: {
    withBorder: 'borderMd',
    text: 'Написать рецензию',
  },
}

export const Sign_In: Story = {
  name: 'Это не кнопка, а ссылка',
  args: {
    icon: 'profile',
    background: 'transparent',
    withBorder: 'borderBg',
    size: 'big',
  },
}

export const Filter: Story = {
  args: {
    background: 'transparent',
    text: 'Фильтры',
  },
}

export const Language: Story = {
  args: {
    icon: 'language',
    background: 'transparent',
    size: 'big',
  },
}
