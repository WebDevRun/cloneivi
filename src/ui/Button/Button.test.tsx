import { composeStories } from '@storybook/react'
import { render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'

import * as stories from './Button.stories'

const { Player, Primary, Secondary } = composeStories(stories)

describe('Button tests', () => {
  it('снимок кнопки Player', () => {
    const tree = renderer.create(<Player />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('отображает кнопку Primary', () => {
    render(<Primary />)
    const buttonElement = screen.getByText(/Оплатить подписку/i)
    expect(buttonElement).not.toBeNull()
  })

  it('отображает кнопку Secondary', () => {
    render(<Secondary />)
    const buttonElement = screen.getByText(/Смотреть по подписке/i)
    expect(buttonElement).not.toBeNull()
  })
})
