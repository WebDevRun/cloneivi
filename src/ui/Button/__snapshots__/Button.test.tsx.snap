// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button tests снимок кнопки Player 1`] = `
<button
  className="button red active borderNone big fitContent"
  type="button"
>
  <div
    className="mainText"
  >
    Смотреть с рекламой
    <div
      className="subText"
    >
      серия 1 сезон 1
    </div>
  </div>
</button>
`;
